### PR TITLE
Count fix and Type alias refactor in Streams Scala API

### PR DESCRIPTION
**Pull request by joan38**
Provide a tiny helper for creating Materialized with implicit Serdes.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by joan38** [comment]
@guozhangwang @debasishg Any thoughts on this?

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedStream.scala [hunk 1]
```diff
@@ -86,7 +85,7 @@ class KGroupedStream[K, V](val inner: KGroupedStreamJ[K, V]) {
    * @see `org.apache.kafka.streams.kstream.KGroupedStream#reduce`
    */ 
   def reduce(reducer: (V, V) => V,
-    materialized: Materialized[K, V, ByteArrayKeyValueStore]): KTable[K, V] = {
+    materialized: Materialized[K, V, KeyValueStore[Bytes, Array[Byte]]]): KTable[K, V] = {
```
Comment (guozhangwang) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedStream.scala]: I think `ByteArrayKeyValueStore` is more succinct than `KeyValueStore[Bytes, Array[Byte]]`? cc @debasishg @seglo for a second thought.

Comment (debasishg) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedStream.scala]: +1 @guozhangwang .. exactly for this reason the type synonym was introduced

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/Materialized.scala [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ */
+package org.apache.kafka.streams.scala
+
+import org.apache.kafka.common.serialization.Serde
+import org.apache.kafka.common.utils.Bytes
+import org.apache.kafka.streams.kstream.{Materialized => JMaterialized}
+import org.apache.kafka.streams.processor.StateStore
+import org.apache.kafka.streams.state._
+
+object Materialized {
```
Comment (debasishg) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/Materialized.scala]: I would go defensive with implicits on `Materialized`. First, this abstraction doesn't add much value since it returns the Java abstraction. Also if someone doesn't wish to pass a serde to a `Materialized`, and use `Materialized.as` they can be inadvertently passed if implicits are in scope. This will lead to errors that are difficult to debug.

**Comment by seglo** [comment]
I don't believe there's much purpose to abstract `Materialized` to help with implicit SerDes now that the internal changes to propagate the SerDes for each of these aggregation operators has been completed in #4919.  As @guozhangwang mentioned in that PR:

> This is a meta explanation about the serde inheritance across multiple internal impl classes:
> 
> reduce: inherit the key and value serdes from the parent XXImpl class.
> count: inherit the key serdes, enforce setting the Serdes.Long() for value serdes.
> aggregate: inherit the key serdes, do not set for value serdes internally (line 92 here is for this case).

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedStream.scala (hunk 1 above)
Comment (joan38) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedStream.scala]: I'm good with the type alias `ByteArrayKeyValueStore ` but then we should do it consistently for  the other types like `SessionStore[Bytes, Array[Byte]]`. I will then modify the PR to add the others.

**Comment by joan38** [comment]
Indeed this doesn't bring too much value so I removed the add of the Materialized helper and left only some refactorings

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedStream.scala [hunk 3]
```diff
@@ -58,10 +58,8 @@ class KGroupedStream[K, V](val inner: KGroupedStreamJ[K, V]) {
    * represent the latest (rolling) count (i.e., number of records) for each key
    * @see `org.apache.kafka.streams.kstream.KGroupedStream#count`
    */ 
-  def count(materialized: Materialized[K, Long, ByteArrayKeyValueStore]): KTable[K, Long] = { 
-    val c: KTable[K, java.lang.Long] = inner.count(materialized.asInstanceOf[Materialized[K, java.lang.Long, ByteArrayKeyValueStore]])
-    c.mapValues[Long](Long2long _)
-  }
+  def count(materialized: Materialized[K, Long, ByteArrayKeyValueStore]): KTable[K, Long] =
+    inner.count(materialized)
```
Comment (guozhangwang) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedStream.scala]: I'm just curious why we can save the casting of `Long2long` now?

Comment (joan38) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedStream.scala]: Not sure why it's been done in the first place

Comment (guozhangwang) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedStream.scala]: We need to double check if Scala 2.11 is compatible with java / scala Long (I'm not an expert in this case..)

If that is indeed the case, could the above `count()` overloaded function be updated as well?

Comment (joan38) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedStream.scala]: There shouldn't be any difference between 2.12 and 2.11 here. Both are just doing boxing and unboxing in Scala long.

I'll write a test to but that on clear tomorrow.

Comment (joan38) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedStream.scala]: Indeed this change is faulty. I reverted this and added a test to make sure we don't get in this in the future.

Comment (debasishg) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedStream.scala]: Actually the other overload `def count(): KTable[K, Long]` doesn't compile without the `mapValues`. This tricked us into thinking that we need to do this for this overload as well. Possibly the type parameter `Long` in `Materialized` helps the compiler to infer the return type of `Long` and make the conversion. So we still need to do the conversion manually in the other overload of `count`.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/TimeWindowedKStream.scala [hunk 4]
```diff
@@ -89,11 +83,8 @@ class TimeWindowedKStream[K, V](val inner: TimeWindowedKStreamJ[K, V]) {
    * represent the latest (rolling) count (i.e., number of records) for each key
    * @see `org.apache.kafka.streams.kstream.TimeWindowedKStream#count`
    */ 
-  def count(materialized: Materialized[K, Long, WindowStore[Bytes, Array[Byte]]]): KTable[Windowed[K], Long] = {
-    val c: KTable[Windowed[K], java.lang.Long] = 
-      inner.count(materialized.asInstanceOf[Materialized[K, java.lang.Long, WindowStore[Bytes, Array[Byte]]]])
-    c.mapValues[Long](Long2long(_))
-  }
+  def count(materialized: Materialized[K, Long, ByteArrayWindowStore]): KTable[Windowed[K], Long] =
+    inner.count(materialized)
```
Comment (guozhangwang) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/TimeWindowedKStream.scala]: Should we revert the change here as well?

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/SessionWindowedKStream.scala [hunk 5]
```diff
@@ -81,10 +74,7 @@ class SessionWindowedKStream[K, V](val inner: SessionWindowedKStreamJ[K, V]) {
    * that represent the latest (rolling) count (i.e., number of records) for each key within a window
    * @see `org.apache.kafka.streams.kstream.SessionWindowedKStream#count`
    */
-  def count(): KTable[Windowed[K], Long] = {
-    val c: KTable[Windowed[K], java.lang.Long] = inner.count()
-    c.mapValues[Long](Long2long(_))
-  }
+  def count(): KTable[Windowed[K], Long] = inner.count().asInstanceOf[KTable[Windowed[K], Long]]
```
Comment (guozhangwang) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/SessionWindowedKStream.scala]: Ditto here.

**Review by guozhangwang — COMMENTED** [review]
Other parts LGTM. Left two more minor comments.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/TimeWindowedKStream.scala (hunk 4 above)
Comment (joan38) [reply streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/TimeWindowedKStream.scala]: Indeed we should

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedTable.scala [hunk 6]
```diff
@@ -56,9 +56,12 @@ class KGroupedTable[K, V](inner: KGroupedTableJ[K, V]) {
    * @return a [[KTable]] that contains "update" records with unmodified keys and `Long` values that
    * represent the latest (rolling) count (i.e., number of records) for each key
    * @see `org.apache.kafka.streams.kstream.KGroupedTable#count`
-   */ 
-  def count(materialized: Materialized[K, Long, ByteArrayKeyValueStore]): KTable[K, Long] =
-    inner.count(materialized)
```
Comment (joan38) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/KGroupedTable.scala:61]: This one would have triggered a StackOverflow looping on itself.

File: streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/SessionWindowedKStream.scala [hunk 7]
```diff
@@ -94,8 +87,11 @@ class SessionWindowedKStream[K, V](val inner: SessionWindowedKStreamJ[K, V]) {
    * that represent the latest (rolling) count (i.e., number of records) for each key within a window
    * @see `org.apache.kafka.streams.kstream.SessionWindowedKStream#count`
    */
-  def count(materialized: Materialized[K, Long, SessionStore[Bytes, Array[Byte]]]): KTable[Windowed[K], Long] =
-    inner.count(materialized)
```
Comment (joan38) [inline streams/streams-scala/src/main/scala/org/apache/kafka/streams/scala/kstream/SessionWindowedKStream.scala:98]: This one would have triggered a StackOverflow looping on itself.

**Comment by guozhangwang** [comment]
Thanks @joan38 ! Merged to trunk.

**Comment by joan38** [comment]
Thanks @guozhangwang
